pkg/helm/urlfetcher.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helm
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mholt/archiver"
+
+	"istio.io/pkg/log"
+)
+
+// FileDownloader is wrapper of HTTP client to download files
+type FileDownloader struct {
+	// client is a HTTP/HTTPS client.
+	client *http.Client
+}
+
+// URLFetcher is used to fetch and manipulate charts from remote url
+type URLFetcher struct {
+	// url is url to download the charts
+	url string
+	// verifyURL is url to download the verification file
+	verifyURL string
+	// verify indicates whether the downloaded tar should be verified
+	verify bool
+	// destDir is path of charts downloaded to, empty as default to temp dir
+	destDir string
+	// untarDir is destination of untar
+	untarDir string
+	// downloader downloads files from remote url
+	downloader *FileDownloader
+}
+
+// fetchChart fetches the charts and verifies charts against shaF if required
+func (f *URLFetcher) fetchChart(shaF string) error {
+	c := f.downloader
+
+	// fetch to default temp dir
+	saved, err := c.DownloadTo(f.url, f.destDir)
+	if err != nil {
+		return err
+	}
+	file, err := os.Open(saved)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if f.verify {
+		// verify with sha file
+		hashAll, err := ioutil.ReadFile(shaF)
+		if err != nil {
+			return fmt.Errorf("failed to read sha file: %s", err)
+		}
+		hash := string(hashAll)
+		h := sha256.New()
+		if _, err := io.Copy(h, file); err != nil {
+			log.Error(err.Error())
+		}
+		sum := h.Sum(nil)
+		if !strings.EqualFold(hex.EncodeToString(sum), hash) {
+			return errors.New("checksum does not match")
+		}
+	}
+
+	// After verification, untar the chart into the requested directory.
+	return archiver.Unarchive(saved, f.untarDir)
+}
+
+// fetch sha file
```

Comment:
Optional: lint will complain about the comment format if the symbol is exported, but I recommend following the same pattern for most unexported symbols, especially functions. It leads to a more consistent style and is more idiomatic.

Reply:
done